### PR TITLE
MF-1159 - add gateway metadata update in provision method

### LINKED ISSUE #1159: Enable provision service to create gateway configuration matching UI gateway
When we create gateway from UI we are setting few properties in `metadata` of gateway `thing` that [Mainflux UI](https://github.com/mainflux/ui) depends on to display the data and work properly.
Provision method should be able to produce same configuration.

### PR DESCRIPTION
**Pull request by mteodor**
Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

Resolves #1159

### CODE REVIEW

File: provision/service.go [hunk 1]
```diff
@@ -67,7 +75,7 @@ func New(cfg Config, sdk SDK.SDK, logger logger.Logger) Service {
 }
 
 // Provision is provision method for adding devices to proxy.
-func (ps *provisionService) Provision(token, externalID, externalKey string) (res Result, err error) {
+func (ps *provisionService) Provision(token, extID, externalKey string) (res Result, err error) {
```
Comment (manuio) [inline provision/service.go]: `externalID` was fine

File: provision/service.go [hunk 2]
```diff
@@ -145,29 +153,36 @@ func (ps *provisionService) Provision(token, externalID, externalKey string) (re
...
 
 	var cert SDK.Cert
+	var bs SDK.BootstrapConfig
 	for _, thing := range things {
 		bootstrap := false
-		if _, ok := thing.Metadata[ExternalID]; ok {
+		if _, ok := thing.Metadata[externalID]; ok {
 			bootstrap = true
 		}
 		chanIDs := []string{}
```
Comment (manuio) [inline provision/service.go]: We should use `var chanIDs ...` here

File: provision/service.go [hunk 3]
```diff
@@ -192,11 +207,49 @@ func (ps *provisionService) Provision(token, externalID, externalKey string) (re
...
 		}
+
 	}
 
+	ps.updateGateway(token, bs, channels)
 	return res, nil
 }
 
+func (ps *provisionService) updateGateway(token string, bs SDK.BootstrapConfig, channels []SDK.Channel) error {
+	g := Gateway{}
```
Comment (manuio) [inline provision/service.go]: `var g Gateway`

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/service.go (hunk 1 above)
Comment (mteodor) [reply provision/service.go]: it overlaps with constant `externalID`

Comment (manuio) [reply provision/service.go]: Change the constant to `extIDKey` or something similar

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/service.go [hunk 4]
```diff
@@ -192,11 +207,49 @@ func (ps *provisionService) Provision(token, externalID, externalKey string) (re
...
+			g.CtrlChannelID = ch.ID
+		case data:
+			g.DataChannelID = ch.ID
+		case export:
+			g.ExportChannelID = ch.ID
+		}
+	}
+	g.GwPassword = bs.ExternalKey
+	g.MAC = bs.ExternalID
+	// same value, keep external id in meta for compatibility
```
Comment (drasko) [inline provision/service.go]: I did not quite understand this comment. Also, they start with capital letters.

**Review by drasko — CHANGES_REQUESTED** [review]

File: provision/service.go (hunk 4 above)
Comment (mteodor) [reply provision/service.go]: this  `MAC` property will be removed, actually

Comment (drasko) [reply provision/service.go]: Why not removing it now then?

File: sdk/go/bootstrap.go [hunk 5]
```diff
@@ -18,12 +18,12 @@ const configsEndpoint = "configs"
...
 const whitelistEndpoint = "state"
 
-// BoostrapConfig represents Configuration entity. It wraps information about external entity
+// BootstrapConfig represents Configuration entity. It wraps information about external entity
 // as well as info about corresponding Mainflux entities.
 // MFThing represents corresponding Mainflux Thing ID.
 // MFKey is key of corresponding Mainflux Thing.
 // MFChannels is a list of Mainflux Channels corresponding Mainflux Thing connects to.
-type BoostrapConfig struct {
+type BootstrapConfig struct {
```
Comment (manuio) [inline sdk/go/bootstrap.go:26]: Good one! You can also fix `UpdateBoostrap` etc etc

**Review by manuio — CHANGES_REQUESTED** [review]

File: sdk/go/sdk.go [hunk 6]
```diff
@@ -175,22 +175,22 @@ type SDK interface {
 	Version() (string, error)
 
 	// AddBootstrap add boostrap configuration
-	AddBootstrap(token string, cfg BoostrapConfig) (string, error)
+	AddBootstrap(token string, cfg BootstrapConfig) (string, error)
 
 	// View returns Thing Config with given ID belonging to the user identified by the given token.
-	ViewBoostrap(token, id string) (BoostrapConfig, error)
+	ViewBoostrap(token, id string) (BootstrapConfig, error)
```
Comment (manuio) [inline sdk/go/sdk.go]: `ViewBoostrap`

File: sdk/go/sdk.go [hunk 7]
```diff
@@ -175,22 +175,22 @@ type SDK interface {
...
 	// Update updates editable fields of the provided Config.
-	UpdateBoostrap(token string, cfg BoostrapConfig) error
+	UpdateBoostrap(token string, cfg BootstrapConfig) error
 
 	// Remove removes Config with specified token that belongs to the user identified by the given token.
 	RemoveBoostrap(token, id string) error
 
 	// Bootstrap returns Config to the Thing with provided external ID using external key.
-	Boostrap(externalKey, externalID string) (BoostrapConfig, error)
+	Boostrap(externalKey, externalID string) (BootstrapConfig, error)
```
Comment (manuio) [inline sdk/go/sdk.go]: `Boostrap`

File: sdk/go/sdk.go [hunk 8]
```diff
@@ -175,22 +175,22 @@ type SDK interface {
...
-	AddBootstrap(token string, cfg BoostrapConfig) (string, error)
+	AddBootstrap(token string, cfg BootstrapConfig) (string, error)
 
 	// View returns Thing Config with given ID belonging to the user identified by the given token.
-	ViewBoostrap(token, id string) (BoostrapConfig, error)
+	ViewBoostrap(token, id string) (BootstrapConfig, error)
 
 	// Update updates editable fields of the provided Config.
-	UpdateBoostrap(token string, cfg BoostrapConfig) error
+	UpdateBoostrap(token string, cfg BootstrapConfig) error
```
Comment (manuio) [inline sdk/go/sdk.go]: `UpdateBoostrap`

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/README.md [hunk 9]
```diff
@@ -120,6 +120,12 @@ In the case that provision service is not deployed with credentials or API key o
 curl -s -S  -X POST  http://localhost:8091/mapping -H "Authorization: <token|api_key>" -H 'Content-Type: application/json' -d '{ "external_id" : "<external_id>", "external_key":"<external_key>" }'
 ```
 
+Or if you want to specify a name for thing different than in `config.toml` you can specify post data as:
+
+```json
+{ "name":"<name>","external_id" : "<external_id>", "external_key":"<external_key>" }
```
Comment (manuio) [inline provision/README.md]: Add white space only after `:` and `,`: `{"name": "<name>", "external_id": "<external_id>", "external_key": "<external_key>"}`

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/service.go [hunk 10]
```diff
@@ -25,6 +31,7 @@ var (
 	ErrFailedBootstrapRetrieval = errors.New("failed to retrieve bootstrap")
 	ErrFailedCertCreation       = errors.New("failed to create certificates")
 	ErrFailedBootstrap          = errors.New("failed to create bootstrap config")
+	ErrFailedGatewayUpdate      = errors.New("failed to updated gateway metadata")
```
Comment (manuio) [inline provision/service.go]: I would use `Err...` instead of `ErrFailed...`

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/README.md [hunk 11]
```diff
@@ -120,6 +120,12 @@ In the case that provision service is not deployed with credentials or API key o
 curl -s -S  -X POST  http://localhost:8091/mapping -H "Authorization: <token|api_key>" -H 'Content-Type: application/json' -d '{ "external_id" : "<external_id>", "external_key":"<external_key>" }'
```
Comment (manuio) [inline provision/README.md]: Add white space only after `:` and `,`: `{"external_id": "<external_id>", "external_key": "<external_key>"}`

**Review by manuio — CHANGES_REQUESTED** [review]

File: provision/README.md (hunk 9 above)
Comment (drasko) [reply provision/README.md]: Yes, this needs uniform representation.

**Review by drasko — CHANGES_REQUESTED** [review]

File: provision/service.go [hunk 12]
```diff
@@ -33,11 +40,11 @@ var _ Service = (*provisionService)(nil)
 type Service interface {
 	// Provision is the only method this API specifies. Depending on the configuration,
 	// the following actions will can be executed:
-	// - create a Thing based od mac address
+	// - create a Thing based on external_id ( eg. MAC address)
```
Comment (manuio) [inline provision/service.go]: `(eg. MAC address)`

**Review by manuio — CHANGES_REQUESTED** [review]

**Review by drasko — DISMISSED** [review]
LGTM

File: provision/service.go [hunk 13]
```diff
@@ -66,10 +73,11 @@ func New(cfg Config, sdk SDK.SDK, logger logger.Logger) Service {
 	}
 }
 
-// Provision is provision method for adding devices to proxy.
-func (ps *provisionService) Provision(token, externalID, externalKey string) (res Result, err error) {
-	channels := make([]SDK.Channel, 0)
-	things := make([]SDK.Thing, 0)
+// Provision is provision method for creating setup according to
+// provision layout specified in config.toml
+func (ps *provisionService) Provision(name, token, extIDVal, externalKey string) (res Result, err error) {
```
Comment (manuio) [inline provision/service.go]: I would keep `externalID`

**Review by manuio — CHANGES_REQUESTED** [review]

**Review by manuio — APPROVED** [review]
LGTM

**Review by blokovi — APPROVED** [review]

**Pull request merged**